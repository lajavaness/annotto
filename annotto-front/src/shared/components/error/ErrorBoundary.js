import { Button, Result } from 'antd'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

class ErrorBoundary extends Component {
	state = {
		error: null,
	}

	static getDerivedStateFromError(error) {
		return { error }
	}

	componentDidCatch(error, info) {
		console.error('Crash:', error, info)
	}

	render() {
		const { children } = this.props
		const { error } = this.state

		if (error) {
			return (
				<Result
					default
					status="500"
					title="500"
					subTitle="Sorry, something went wrong."
					extra={
						<Button type="link" href="/">
							Back Home
						</Button>
					}
				/>
			)
		}

		return children
	}
}

ErrorBoundary.propTypes = {
	/** Content to display if there are no errors. */
	children: PropTypes.node,
}

ErrorBoundary.defaultProps = {
	children: null,
}

export default ErrorBoundary
