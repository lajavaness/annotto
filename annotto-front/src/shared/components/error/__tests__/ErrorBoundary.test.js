import { act, render } from '@testing-library/react'
import React from 'react'

import ErrorBoundary from '../ErrorBoundary'

const getInstance = (props = {}) => <ErrorBoundary>{props.children}</ErrorBoundary>

const pauseErrorLogging = (codeToRun) => {
	const logger = console.error
	console.error = () => {}
	codeToRun()

	console.error = logger
}

describe('ErrorBoundary', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})

	test('act works in this case', async () => {
		await act(async () => {
			const { getByText } = render(getInstance({ children: <div>ErrorBoundary</div> }))
			expect(getByText('ErrorBoundary')).toBeVisible()
		})
	})

	it('renders children', () => {
		const { getByText } = render(getInstance({ children: <div>ErrorBoundary</div> }))
		expect(getByText('ErrorBoundary')).toBeVisible()
	})

	it('renders error message', () => {
		const Child = () => {
			throw new Error('Errored!')
		}

		pauseErrorLogging(() => {
			const { getByText } = render(getInstance({ children: <Child /> }))

			expect(getByText('Sorry, something went wrong.')).toBeVisible()
		})
	})
})
