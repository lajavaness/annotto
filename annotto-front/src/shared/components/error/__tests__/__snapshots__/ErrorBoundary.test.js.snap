// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary matches snapshot 1`] = `<DocumentFragment />`;
